File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/infrastructure/zlink/handlers/AllocateConversationHandler.java
package systems.zlink.samples.supportchat.server.support.infrastructure.zlink.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.supportchat.server.support.application.assignment.SupportConversationAllocator;
import systems.zlink.samples.supportchat.server.support.domain.conversation.ConversationModels.Statuses;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

@ZLinkHandlerGroup("support")
public final class AllocateConversationHandler
    implements ZLinkRequestHandler<
        Messages.AllocateConversationReq,
        Messages.AllocateConversationRes> {
    private final SupportConversationAllocator allocator;

    public AllocateConversationHandler(SupportConversationAllocator allocator) {
        this.allocator = allocator;
    }

    @Override
    public Messages.AllocateConversationRes handle(
        Messages.AllocateConversationReq request,
        ZLinkRequestContext context) {
        String conversationId = allocator.allocate(
            request.customerActorId(),
            request.customerDisplayName(),
            request.subject());
        return new Messages.AllocateConversationRes(conversationId, Statuses.WaitingForAgent);
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/adapters/zlink/handlers/AllocateConversationHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.adapters.zlink.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.supportchat.server.configuration.ConversationStatuses
import systems.zlink.samples.kotlin.supportchat.server.support.application.assignment.SupportConversationAllocator
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AllocateConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AllocateConversationRes

@ZLinkHandlerGroup("support")
class AllocateConversationHandler(
    private val allocator: SupportConversationAllocator,
) : ZLinkSuspendingRequestHandler<AllocateConversationReq, AllocateConversationRes> {
    override suspend fun handle(
        request: AllocateConversationReq,
        context: ZLinkRequestContext,
    ) = run {
        val conversationId = allocator.allocate(
            request.customerActorId,
            request.customerDisplayName,
            request.subject,
        )
        AllocateConversationRes(conversationId, ConversationStatuses.WaitingForAgent)
    }
}
